Maple2.Server.Game/PacketHandlers/FishingHandler.cs

```diff
@@ -0,0 +1,377 @@
+                break;
+            case Command.Catch:
+                HandleCatch(session, packet);
+                break;
+            case Command.Start:
+                HandleStart(session, packet);
+                break;
+            case Command.FailMinigame:
+                HandleFailMinigame(session);
+                break;
+        }
+    }
+
+    private void HandlePrepare(GameSession session, IByteReader packet) {
+        long fishingRodUid = packet.ReadLong();
+
+        if (session.Field == null || session.GuideObject != null) {
+            return;
+        }
+
+        Item? rod = session.Item.Inventory.Get(fishingRodUid);
+        if (rod == null || rod.Metadata.Function?.Type != ItemFunction.FishingRod) {
+            session.Send(FishingPacket.Error(FishingError.s_fishing_error_invalid_item));
+            return;
+        }
+
+        if (!int.TryParse(rod.Metadata.Function?.Parameters, out int rodCode) || !TableMetadata.FishingRodTable.Entries.TryGetValue(rodCode, out FishingRodTable.Entry? rodMetadata)) {
+            session.Send(FishingPacket.Error(FishingError.s_fishing_error_invalid_item));
+            return;
+        }
+
+        if (session.Mastery[MasteryType.Fishing] < rodMetadata.MinMastery) {
+            session.Send(FishingPacket.Error(FishingError.s_fishing_error_fishingrod_mastery));
+            return;
+        }
+
+        if (!TableMetadata.FishingSpotTable.Entries.TryGetValue(session.Player.Value.Character.MapId, out FishingSpotTable.Entry? spotMetadata)) {
+            session.Send(FishingPacket.Error(FishingError.s_fishing_error_notexist_fish));
+            return;
+        }
+
+        if (session.Mastery[MasteryType.Fishing] < spotMetadata.MinMastery) {
+            session.Send(FishingPacket.Error(FishingError.s_fishing_error_lack_mastery));
+            return;
+        }
+
+        IList<Vector3> fishingBlocks = GetFishingBlocks(session.Player.Position, session.Player.Rotation);
+
+        session.GuideObject = session.Field?.SpawnGuideObject(session.Player, new FishingGuideObject(), GetGuideObjectPosition(fishingBlocks, session.Player.Position));
+        session.GuideObject!.RodMetadata = rodMetadata;
+
+        session.Field?.Broadcast(GuideObjectPacket.Create(session.GuideObject));
+        session.Send(FishingPacket.LoadTiles(fishingBlocks));
+        session.Send(FishingPacket.Prepare(fishingRodUid));
+    }
+
+    private static Vector3 GetGuideObjectPosition(IEnumerable<Vector3> fishingBlocks, Vector3 playerPosition) {
+        // get the fishing block closest to player
+        Vector3 guidePosition = fishingBlocks.MinBy(o => Math.Sqrt(Math.Pow(playerPosition.X - o.X, 2) + Math.Pow(playerPosition.Y - o.Y, 2)));
+
+        // Move guide 1 block up so it sits on top of the fishing block
+        guidePosition.Z += 150;
+        return guidePosition;
+    }
+
+    private static IList<Vector3> GetFishingBlocks(Vector3 position, Vector3 rotation) {
```

Comment:
rather than building up a list, you can just make this return `IEnumerable<Vector3>` and do `yield return checkBlock` for every block you would have added to the list.

For usage of this function you would do `GetFishingBlocks(position, rotation).ToArray()`